security/security.tex

```diff
@@ -0,0 +1,331 @@
+
+Security isn't just making sure that a program can be manipulated to your ends.
+Some times, security is making sure that the program can't be crashed or effectively timed with input.
+An example of the former is any attack where the attacker threatens to shut a system down due to a leak.
+This means that the system has a flaw where the user can input a value and cause the system to fail.
+One can imagine in mission critical systems -- power grids, medical devices, etc -- this isn't a lightly taken threat.
+
+In addition, a lot of novice programmers neglect trying the make a program look similar given all inputs.
+A common example of this is comparing two strings.
+Let's say that you are guessing the CSRF (cross site request forgery) token on a website.
+If the programmed server returns the request immediately after the token doesn't match the server's, that is a security bug.
+If the token matches none of the characters, an attacker now know that a quick response means that the first few characters were off.
+If the token matches some of the characters but not all, an attacker know that this request will take slightly longer.
+It is important to balance speed an security.
+We want to use a fast \keyword{memcmp} but it may not be secure.
+
+\section{Cyber Security}
+
+Cyber Security is arguably the most popular area of security.
+More and more of our systems are hacked over the web, it is important to understand how we can protect against these attacks
+
+\subsection{Security at the TCP Level}
+
+\begin{enumerate}
+\item Encryption. \textbf{TCP is not encrypted!} This means any data that is sent over a TCP connection is in plain text. If one needs to send encrypted data, one needs to use a higher level protocol such as HTTPs or develop their own.
+\item Identity Verification. In TCP, there is no way to verify the identity of who the program is connecting to. There are no checks or federated databases in place. One just has to trust the the DNS server gave a resonable response which is almost always the incorrect answer. Apart from systems that have a approved white list or a ``secret'' connection protocol, there is little at the TCP level that one can do to stop.
+\item Syn-Ack Sequence Number. This is a security improvement. TCP features what we call sequence numbers. That means that during the SYN-SYN/ACK-ACK dance, a connection starts at a random integer. This is important because if an attacker is trying to spoof packets (pretend those packets are coming from your program) that means that the attacker must either correctly guess -- which is hard -- or be in the route that your packet takes to the destination -- much more likely. ISPs help out with the destination problem because it may send a connection through varying routers which makes it hard for an attacker to sit anywhere and be sure that they will receive your packets -- this is why security experts usually advise against using coffee shop wifi for sensitive tasks.
```

Comment:
```suggestion
\item TCP Sequence Numbers. 
This is a security improvement. 
TCP features what we call sequence numbers. 
During the SYN, SYN/ACK, ACK dance, both sides of the connection start with at a random integer called the sequence number. 
This is important, because if an attacker is trying to spoof packets (pretend those packets are coming from your program) that means that the attacker must either correctly guess a number, which must be within a narrow window -- which is hard -- or be in the route that your packet takes to the destination -- much more likely. 
ISPs help out with the destination problem because it may send a connection through varying routers which makes it hard for an attacker to sit anywhere and be sure that they will receive your packets -- this is why security experts usually advise against using coffee shop WiFi (more specifically, public unsecured networks) for sensitive tasks.
```